src/vivarium_cluster_tools/runner.py

```diff
@@ -229,90 +230,225 @@ def build_job_list(ctx):
-                        retries += 1
-                        sleep(backoff)
-                end = time()
-                logger.info(f'\t\tfetched job in {end - start:.4f}')
-                start = end
-                result = pd.read_msgpack(job.result[0])
-                end = time()
-                logger.info(f'\t\tread from msgpack in {end - start:.4f}')
-                chunk_results.append(result)
-                dirty = True
-                finished_registry.remove(job)
+                result = self.get_result(registry_key, job_id)
+                if result is not None:
+                    chunk_results.append(result)
+            yield chunk_results
+
+    def get_job(self, registry_key, job_id):
+        logger.info(f'Fetching job {job_id} from queue {registry_key}.')
 
-            if dirty:
+        job = None
+        while self._retries[registry_key] > 0:
+            try:
                 start = time()
-                results = pd.concat([results] + chunk_results, axis=0)
+                job = self._queues[registry_key].fetch_job(job_id)
                 end = time()
-                logger.info(f"\t\tConcatenated batch in {end - start:.4f}")
+                logger.info(f'Fetched job {job_id} from queue {registry_key} in {end - start:.2f}s')
+                break
+            except redis.exceptions.ConnectionError:
+                self.sleep_on_it(registry_key)
+            except KeyError:
+                logger.exception(f'Queue {registry_key} already complete or failed.')
+                break
+
+        if job is None:
+            self.abandon_queue(registry_key)
+
+        return job
+
+    def get_result(self, registry_key, job_id):
+        job = self.get_job(registry_key, job_id)
+        result = None
+        if job is not None:
+            start = time()
+            result = pd.read_msgpack(job.result[0])
+            end = time()
+            logger.info(f'Read {job_id} from msgpack from queue {registry_key} in {end - start:.2f}s.')
+            self._finished[registry_key].remove(job)
+            self._status[registry_key]['finished'] += 1
+        return result
+
+    def sleep_on_it(self, registry_key):
+        backoff = RegistryManager.backoff * random.random()
+        logger.warning(f'Failed to connect to redis for queue {registry_key}.  Retrying in {backoff}s.')
+        self._retries[registry_key] -= 1
+        sleep(backoff)
+
+    def abandon_queue(self, registry_key):
+        logger.error(f'Lost connection to redis for queue {registry_key}.  Abandoning redis instance.')
+        self._status[registry_key]['failed'] += self._status[registry_key]['pending'] + self._status[registry_key]['running']
+        self._status[registry_key]['pending'] = 0
+        self._status[registry_key]['running'] = 0
+        # TODO: Probably should cleanup, but not sure how yet.
+        try:
+            self._failed.append(
```

Comment:
why do you keep this _failed and _completed lists? do you do anything with them?